pageserver/benches/bench_walredo.rs

```diff
@@ -74,114 +75,83 @@ fn redo_scenarios(c: &mut Criterion) {
-
-        let jhs = (0..threads)
-            .map(|_| {
-                std::thread::spawn({
-                    let manager = manager.clone();
-                    let barrier = barrier.clone();
-                    let work_rx = work_rx.clone();
-                    move || {
-                        let rt = tokio::runtime::Builder::new_current_thread()
-                            .enable_all()
-                            .build()
-                            .unwrap();
-                        let handle = rt.handle();
-                        loop {
-                            // queue up and wait if we want to go another round
-                            if work_rx.lock().unwrap().recv().is_err() {
-                                break;
-                            }
-
-                            let input = Some(input_factory());
-
-                            barrier.wait();
-
-                            execute_all(input, handle, &manager).unwrap();
-
-                            barrier.wait();
-                        }
+    assert_ne!(requesters, 0);
+
+    let rt = tokio::runtime::Builder::new_multi_thread()
+        .enable_all()
+        .build()
+        .unwrap();
+
+    let cancel = CancellationToken::new();
+
+    let barrier = Arc::new(tokio::sync::Barrier::new(requesters + 1));
+
+    let jhs = (0..requesters)
+        .map(|_| {
+            let manager = manager.clone();
+            let barrier = barrier.clone();
+            let cancel = cancel.clone();
+            rt.spawn(async move {
+                let work_loop = async move {
+                    loop {
+                        let input = input_factory();
+                        barrier.wait().await;
+                        let page = input.execute(&manager).await.unwrap();
+                        assert_eq!(page.remaining(), 8192);
+                        barrier.wait().await;
                     }
-                })
-            })
-            .collect::<Vec<_>>();
-
-        let _jhs = JoinOnDrop(jhs);
-
-        b.iter_batched(
-            || {
-                for _ in 0..threads {
-                    work_tx.send(()).unwrap()
+                };
+                tokio::select! {
+                    _ = work_loop => {},
+                    _ = cancel.cancelled() => { }
                 }
-            },
-            |()| {
-                // start the work
-                barrier.wait();
-
+            })
+        })
+        .collect::<Vec<_>>();
+
+    b.iter_batched(
+        || (),
+        |()| {
+            // start the work
+            rt.block_on(async {
+                barrier.wait().await;
                 // wait for work to complete
-                barrier.wait();
-            },
-            criterion::BatchSize::PerIteration,
-        );
+                barrier.wait().await;
+            });
+        },
+        criterion::BatchSize::PerIteration,
+    );
 
-        drop(work_tx);
-    }
-}
+    cancel.cancel();
```

Comment:
So as far as I remember this benchy, the requests were done simultaneously but now it has turned to do as many as fast as possible and for that we need the cancellation token.

Reply:
I vaguely remember that we did not end up with a cancellation safe walredo manager, as such this doesn't sense to use cancellation tokens like this; the tasks could just be descheduled by dropping a joinset. Still it feels this may have some further issues. 

Manager probably should be dropped with the runtime right?

Also there's no prewarming of the process it would seem.

Reply:
Let me sure I understand your concerns correctly:

1. You're pointing out that the walredo manager isn't written to be cancellation safe, which this PR's changes require.
2. You are worried about cleanup ("dropped with the runtime"), I guess, specifically, the cleanup of the child process.
3. The pre-warming is gone.

I'll address (3) right away, please confirm that (1) and (2) exhaustively summarize your comments.

Reply:
Regarding (1) and (2)

I'm wondering: maybe we are cancellation safe now.

```
    /// # Cancel-Safety
    ///
    /// This method is cancellation-safe.
    pub async fn request_redo(
```

I think I audited the code for that in the past.
And actually there's not a single .await inside
Will have to think about this in the async walredo PR (#6628 ) there though, added a TODO there.

Reply:
With the joinset the cancellation token is is not needed; if we are going to cancel by drop we can just `drop(joinset);`.